Custom pydantic conversions

Matt343 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Add support for overriding the `from_pydantic` and `to_pydantic` methods on Pydantic converted types. This allows users to have complete control over the behavior of the conversions if needed, while still preserving the type references that we set up when you register a Pydantic type.

I also fixed up the behavior for when you use a field with the same name as one of the Pydantic fields but use a different type. When you do this, we call the constructor of the type annotation on the value we get from Pydantic before returning it in a query. This is a handy shortcut for using scalar types in Pydantic models or handling fields like `Union[int, str]` in a quick way. The problem before was that it was sort of a happy accident that this worked at all, and it didn't respect aliases and didn't work on objects with only one field. I fixed this behavior so it is more consistent in these cases and added documentation so this is an expected feature now.

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [x] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

* Fixes #1617 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

botberry commented:
Thanks for adding the `RELEASE.md` file!

![](https://media.giphy.com/media/xq1FxHkABwW7m/giphy.gif)

Here's a preview of the changelog:

---

Add better support for custom Pydantic conversion logic and standardize the
behavior when not using `strawberry.auto` as the type.

See https://strawberry.rocks/docs/integrations/pydantic#custom-conversion-logic for details and examples.


---
Here's the preview release card for twitter:

![](https://i.imgur.com/Qyd5SKV.png)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Matt Allen for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

thejaminator commented on tests/experimental/pydantic/test_basic.py:
```diff
@@ -658,3 +659,45 @@ def updateGroup(group: GroupInput) -> GroupOutput:
+
+    assert field1.python_name == "age_"
+    assert field1.type is str
+    assert field1.graphql_name == "age"
```
i don't think this should block your MR. 

but what if i don't want to use the alias in the graphql name?

i was previously providing the type manually so that it wouldn't do that. but i agree its not nice.

we'll handle the alias setting here? https://github.com/strawberry-graphql/strawberry/pull/1546

Matt343 replied:
Yeah I think having a dedicated option to control that is much better

thejaminator commented on docs/integrations/pydantic.md:
```diff
@@ -313,3 +313,119 @@ class Query:
+    content_description: Optional[str] = None
+
+    @staticmethod
+    def from_pydantic(user: User, extra: Dict[str, Any] = None) -> "UserType":
```
i'm not sure whether we should enforce / encourage that the signature should have the keywords "instance", and "extra" 
as specified in the protocol `StrawberryTypeFromPydantic`. 

```suggestion
    def from_pydantic(instance: User, extra: Dict[str, Any] = None) -> "UserType":
```

currently conversion logic wise, it shouldn't affect anything since no code calls from_pydantic. 
but both mypy and pycharm will think that the keyword is "instance" instead of "user" here. 
so people may pass in the wrong keyword.

thejaminator replied:
<img width="935" alt="image" src="https://user-images.githubusercontent.com/30519287/152276909-4a3c56f5-362e-4ac5-bfec-f6bd2d6c9ce4.png">

Matt343 replied:
Fixed the example at least, I'm not sure if we should enforce it though since using another name does technically work aside from IDEs.

thejaminator commented:
@Matt343 theres also the case where the custom conversion is a nested inside another basemodel. 
should the parent call the custom from_pydantic method of the child? (currently it seems like it doesn't)


```python
def test_custom_conversion_functions_nested():
    class User(BaseModel):
        age: int
        password: Optional[str]

    @strawberry.experimental.pydantic.type(User)
    class UserType:
        age: str
        password: strawberry.auto

        @staticmethod
        def from_pydantic(user: User, extra: Dict[str, Any] = None) -> "UserType":
            return UserType(
                age=str(user.age),
                password=base64.b64encode(user.password.encode()).decode()
                if user.password
                else None,
            )

        def to_pydantic(self) -> User:
            return User(
                age=int(self.age),
                password=base64.b64decode(self.password.encode()).decode()
                if self.password
                else None,
            )

    class Parent(BaseModel):
        user: User

    @strawberry.experimental.pydantic.type(Parent)
    class ParentType:
        user: strawberry.auto

    user = User(age=1, password="abc")
    user_strawberry = UserType.from_pydantic(user)

    assert user_strawberry.age == "1"
    assert user_strawberry.password == "YWJj"

    user_pydantic = user_strawberry.to_pydantic()
    assert user == user_pydantic

    parent = Parent(user=user)
    parent_strawberry = ParentType.from_pydantic(parent)
    assert parent_strawberry.user.age == "1" # this fails
    assert parent_strawberry.user.password == "YWJj" # this fails

```

Matt343 commented:
> @Matt343 theres also the case where the custom conversion is a nested inside another basemodel. should the parent call the custom from_pydantic method of the child? (currently it seems like it doesn't)

Good catch, I've added tests for this and revealed a bug in my initial implementation around interfaces. I've changed the method used to override the conversion functions to be more explicit so we can tell the difference between an inherited method and a user supplied one.

thejaminator commented:
> > @Matt343 theres also the case where the custom conversion is a nested inside another basemodel. should the parent call the custom from_pydantic method of the child? (currently it seems like it doesn't)
> 
> Good catch, I've added tests for this and revealed a bug in my initial implementation around interfaces. I've changed the method used to override the conversion functions to be more explicit so we can tell the difference between an inherited method and a user supplied one.

ah, i liked the previous way of just defining the to_pydantic and from_pydantic in the class rather than supplying in the decorator. was the change solely due to not being able to tell the difference? i can try it out

Matt343 commented:
> > > @Matt343 theres also the case where the custom conversion is a nested inside another basemodel. should the parent call the custom from_pydantic method of the child? (currently it seems like it doesn't)
> > 
> > 
> > Good catch, I've added tests for this and revealed a bug in my initial implementation around interfaces. I've changed the method used to override the conversion functions to be more explicit so we can tell the difference between an inherited method and a user supplied one.
> 
> ah, i liked the previous way of just defining the to_pydantic and from_pydantic in the class rather than supplying in the decorator. was the change solely due to not being able to tell the difference? i can try it out

Yeah basically you can't tell if the decorated class is inheriting the methods (in which case we need to redefine them in order for interfaces to work properly) or if they are user defined. Because of that you don't know if you should override them with the default methods or not.

thejaminator commented:
> Yeah basically you can't tell if the decorated class is inheriting the methods (in which case we need to redefine them in order for interfaces to work properly) or if they are user defined. Because of that you don't know if you should override them with the default methods or not.

hmmm yes i'm not sure for interfaces because i think we could just disallow users to declare from_pydantic and to_pydantic on interfaces because it doesn't make sense?
but yes inheritance from another strwaberry type is tricky

Matt343 commented:
> > Yeah basically you can't tell if the decorated class is inheriting the methods (in which case we need to redefine them in order for interfaces to work properly) or if they are user defined. Because of that you don't know if you should override them with the default methods or not.
> 
> hmmm yes i'm not sure for interfaces because i think we could just disallow users to declare from_pydantic and to_pydantic on interfaces because it doesn't make sense? but yes inheritance from another strawberry.type is tricky

I think you could use the `__qualname__` property on the function to tell if it comes from a superclass, bu that seems like a bit of a hack

thejaminator commented:
> > > Yeah basically you can't tell if the decorated class is inheriting the methods (in which case we need to redefine them in order for interfaces to work properly) or if they are user defined. Because of that you don't know if you should override them with the default methods or not.
> > 
> > 
> > hmmm yes i'm not sure for interfaces because i think we could just disallow users to declare from_pydantic and to_pydantic on interfaces because it doesn't make sense? but yes inheritance from another strawberry.type is tricky
> 
> I think you could use the `__qualname__` property on the function to tell if it comes from a superclass, bu that seems like a bit of a hack

if we do that hack, we can raise an exception if the user doesn't redefine the method in the child class. otherwise its not intuitive whether the child class will use the parent's method or the default one. just thinking :)

Matt343 commented:
> if we do that hack, we can raise an exception if the user doesn't redefine the method in the child class. otherwise its not intuitive whether the child class will use the super method or the default one. just thinking :)

Not sure I follow this, you'd want to just detect if the supplied method was defined on the decorated class or a superclass. If it's from a superclass, we can safely override with the default, otherwise we would leave it. Not sure where we would want to raise something.

Overall, that method seems a bit more implicit to me, there's more weird rules about where executed code comes from and that could be confusing.

thejaminator commented:
> Not sure I follow this, you'd want to just detect if the supplied method was defined on the decorated class or a superclass. If it's from a superclass, we can safely override with the default, otherwise we would leave it. Not sure where we would want to raise something.

i think we'll talking about the same thing. I'm suggesting that to avoid the implicitness. we don't override with the default child classes if the parent implements it. we error out at class definition time saying that the parent class has a manual from_pydantic, so the child class needs to have a manual from_pydantic too.

```python
    @strawberry.experimental.pydantic.type(User)
    class UserType:
        age: str
        password: strawberry.auto
    
        @staticmethod
        def from_pydantic(user: User, extra: Dict[str, Any] = None) -> "UserType":
            ...
        
    class AnotherUserType(UserType):
        another_field: str 
        
    AnotherUserType.from_pydantic(something)
    # am i calling the default from_pydantic or the parent UserType's from_pydantic
    # not clear even though we can leave it as the default
    # so we'll raise an error at class definition time that they should manually define it

```

Matt343 commented:
> i think we'll talking about the same thing. I'm suggesting that to avoid the implicitness. we don't override with the default child classes if the parent implements it. we error out at class definition time saying that the parent class has a manual from_pydantic, so the child class needs to have a manual from_pydantic too.
> 
> which i think is reasonable enough. because if the parent has a manual from_pydantic, the child probably needs one too?
> 
> ```python
>     @strawberry.experimental.pydantic.type(User)
>     class UserType:
>         age: str
>         password: strawberry.auto
>     
>         @staticmethod
>         def from_pydantic(user: User, extra: Dict[str, Any] = None) -> "UserType":
>             ...
>         
>     class AnotherUserType(UserType):
>         another_field: str 
>         
>     AnotherUserType.from_pydantic(something)
>     # am i calling the default from_pydantic or the parent UserType's from_pydantic
>     # not clear even though we can leave it as the default
>     # so we'll raise an error at class definition time that they should manually define it
> ```

Ah interesting, the issue I see is that we don't actually know if the parent class method is the default or a custom override, because if you just define a normal interface with no custom method it will still have the methods defined by the time we get to the child class. I guess we could add a property to the default methods that would let us know when we find them...
